Lib/fractions.py

```diff
@@ -247,12 +255,23 @@ def limit_denominator(self, max_denominator=1000000):
             n, d = d, n-a*d
 
         k = (max_denominator-q0)//q1
-        bound1 = Fraction(p0+k*p1, q0+k*q1)
-        bound2 = Fraction(p1, q1)
-        if abs(bound2 - self) <= abs(bound1-self):
-            return bound2
+        bound1_n = p0 + k*p1
+        bound1_d = q0 + k*q1
+        bound2_n = p1
+        bound2_d = q1
+
+        # diff1_n = numerator of (bound1 minus self) as a Fraction;
+        # etc. for diff1_d, diff2_n, diff2_d
+        diff1_n = abs(bound1_n*d_original - n_original*bound1_d)
+        diff1_d = d_original * bound1_d
+        diff2_n = abs(bound2_n*d_original - n_original*bound2_d)
+        diff2_d = d_original * bound2_d
+
+        if diff1_n * diff2_d >= diff2_n * diff1_d:
+            # bound2 is closer (or equal) to original as bound 1
+            return Fraction(bound2_n, bound2_d, _normalize=False)
```

Comment:
I just realised that I'm not 100% convinced that the result here _is_ always normalised. There are two parts to being normalised: the numerator must be relatively prime to the denominator, and the denominator must be positive. I'd thought about the first part (which is fine), but not about the second.

Is it clear that `bound2_d` and `bound1_d` are positive at this point?

Reply:
Answer: yes, it's true that `bound2_d` and `bound1_d` are positive at this point, though perhaps not 100% clear. In the Euclidean loop above, on the very _first_ iteration `a` can be negative, but on all subsequent iterations `a` is nonnegative. And on the first iteration `q1 = 0`, so that potentially negative `a` does no harm, and `q0` and `q1` are guaranteed nonnegative at all times. Then it's easy to check that they must in fact be positive at the point where the loop exits.

Reply:
That was my reading of the loop as well, but I didn't write the code (nor name the variables) so I didn't want to change this without another set of eyes; thanks!